File: android/app/src/main/java/com/aegis/health/tools/GetGuideline.kt
package com.aegis.health.tools

import com.aegis.health.db.KBDatabase
import com.aegis.health.models.GuidelineRecommendation
import kotlinx.serialization.Serializable

@Serializable
data class GetGuidelineResult(
    val recommendations: List<GuidelineRecommendation> = emptyList(),
    val gaps: List<String> = emptyList(),
    val error: String? = null,
)

object GetGuideline {

    fun getGuidelines(
        age: Int,
        sex: String,
        conditions: List<String>?,
        db: KBDatabase,
    ): GetGuidelineResult {
        if (age <= 0) {
            return GetGuidelineResult(error = "Invalid age")
        }

        val normalizedSex = when (sex.lowercase()) {
            "m", "male" -> "male"
            "f", "female" -> "female"
            else -> "all"
        }

        val rows = db.queryGuidelines(age, normalizedSex, conditions)

        if (rows.isEmpty()) {
            return GetGuidelineResult(
                error = "No USPSTF recommendations found for age=$age, sex=$normalizedSex",
            )
        }

        val recommendations = rows.map { row ->
            GuidelineRecommendation(
                title = row["title"] ?: "",
                grade = row["grade"] ?: "",
                description = row["description"] ?: "",
                population = synthesizePopulation(
                    row["population_age_min"],
                    row["population_age_max"],
                    row["population_sex"],
                ),
                citation = row["source"] ?: "USPSTF",
            )
        }

        val gaps = buildGapsList(age, normalizedSex, conditions)

        return GetGuidelineResult(
            recommendations = recommendations,
            gaps = gaps,
        )
    }

    /**
     * Turn the three population_* columns into a human-readable label such as
     * "Men 65+", "Women 50-74", or "Children 0-5". NULL bounds are treated as
     * "any" on that side.
     */
    private fun synthesizePopulation(ageMin: String?, ageMax: String?, sex: String?): String {
        val min = ageMin?.toIntOrNull()
        val max = ageMax?.toIntOrNull()
        val s = (sex ?: "all").trim().lowercase()
        val pediatric = max != null && max <= 18
        val noun = when {
            s == "male" -> if (pediatric) "Boys" else "Men"
            s == "female" -> if (pediatric) "Girls" else "Women"
            else -> if (pediatric) "Children" else "Adults"
        }
        return when {
            min != null && max != null -> "$noun $min-$max"
            min != null -> "$noun $min+"
            max != null -> "$noun under ${max + 1}"
            else -> "All ${noun.lowercase()}"
        }
    }

    private fun buildGapsList(age: Int, sex: String, conditions: List<String>?): List<String> {
        val gaps = mutableListOf<String>()
        if (conditions.isNullOrEmpty()) {
            gaps += "No conditions provided — condition-specific screenings may be missing."
        }
        if (sex == "all") {
            gaps += "Sex not specified — sex-specific screenings may be missing."
        }
        return gaps
    }
}

File: submission/hf_space/tools/tools/get_guideline.py
"""Retrieve USPSTF preventive-care guidelines matching patient demographics.

Reads against the canonical schema in kb/schema.sql:
  id, recommendation_id, title, grade,
  population_age_min, population_age_max, population_sex,
  description, clinical_url, source

The external contract the LLM receives (fields in each recommendation dict)
is kept stable: {title, grade, description, population, citation}.
- ``population`` is synthesized from the three population_* columns
  (e.g., "Men 65+", "Women 50-74", "Adults 40-75").
- ``citation`` maps to the ``source`` column (e.g., "USPSTF (2024 curated snapshot)").
"""

from __future__ import annotations

import sqlite3
from pathlib import Path

DEFAULT_DB = "kb/output/aegis_kb.sqlite"


def _synthesize_population(age_min: int | None, age_max: int | None, sex: str | None) -> str:
    """Turn the three population_* columns into a human-readable label."""
    s = (sex or "all").strip().lower()
    # Choose a noun that matches sex, adjusting for pediatric ranges.
    pediatric = age_max is not None and age_max <= 18
    if s == "male":
        noun = "Boys" if pediatric else "Men"
    elif s == "female":
        noun = "Girls" if pediatric else "Women"
    else:
        noun = "Children" if pediatric else "Adults"

    if age_min is not None and age_max is not None:
        return f"{noun} {age_min}-{age_max}"
    if age_min is not None:
        return f"{noun} {age_min}+"
    if age_max is not None:
        return f"{noun} under {age_max + 1}"
    return f"All {noun.lower()}"


def _row_to_rec(row: sqlite3.Row) -> dict:
    return {
        "title": row["title"],
        "grade": row["grade"],
        "description": row["description"],
        "population": _synthesize_population(
            row["population_age_min"], row["population_age_max"], row["population_sex"]
        ),
        "citation": row["source"] or "USPSTF",
    }


def get_guideline(
    age: int,
    sex: str,
    conditions: list[str] | None = None,
    db_path: str = DEFAULT_DB,
) -> dict:
    """Query USPSTF guidelines that apply to the given patient profile.

    Matches by age range (inclusive; NULL means "any") and sex (exact match
    or 'all'). When ``conditions`` is provided, also returns any additional
    Grade A/B recommendations whose title or description substring-matches
    any provided condition (case-insensitive).
    """
    try:
        age = int(age)
    except (TypeError, ValueError):
        return {"error": "Age must be an integer"}

    sex = "" if sex is None else str(sex)
    if not sex.strip() or sex.strip().lower() not in ("male", "female", "m", "f"):
        return {"error": "Sex must be one of: male, female, m, f"}

    db = Path(db_path)
    if not db.exists():
        return {"error": f"Knowledge base not found at {db_path}"}

    sex_normalized = sex.strip().lower()
    if sex_normalized == "m":
        sex_normalized = "male"
    elif sex_normalized == "f":
        sex_normalized = "female"

    conditions = [str(c).strip().lower() for c in (conditions or []) if c is not None and str(c).strip()]

    try:
        conn = sqlite3.connect(str(db))
        conn.row_factory = sqlite3.Row
        cur = conn.cursor()

        # Primary match: age + sex. NULL age bounds mean "any age" on that
        # side. Excludes risk-only ACIP rows — those require an explicit
        # condition match in phase 2 below.
        has_risk_only = bool(conn.execute(
            "SELECT 1 FROM pragma_table_info('guidelines') WHERE name = 'risk_only'"
        ).fetchone())
        risk_only_filter = (
            " AND (risk_only IS NULL OR risk_only = 0)" if has_risk_only else ""
        )
        cur.execute(
            f"""
            SELECT title, grade, description,
                   population_age_min, population_age_max, population_sex,
                   source
            FROM guidelines
            WHERE grade IN ('A', 'B', 'I')
              AND (population_age_min IS NULL OR population_age_min <= ?)
              AND (population_age_max IS NULL OR population_age_max >= ?)
              AND (LOWER(COALESCE(population_sex, 'all')) = ?
                   OR LOWER(COALESCE(population_sex, 'all')) = 'all')
              {risk_only_filter}
            ORDER BY grade, title
            """,
            (age, age, sex_normalized),
        )
        recommendations = [_row_to_rec(row) for row in cur.fetchall()]
        seen_titles = {r["title"] for r in recommendations}

        # Secondary match: condition keywords substring-matched against
        # title or description.  Only returns rows not already in the
        # demographic match so we don't duplicate.
        if conditions:
            like_clause = " OR ".join(
                ["LOWER(title) LIKE ? OR LOWER(description) LIKE ?"] * len(conditions)
            )
            params: list = []
            for term in conditions:
                pat = f"%{term}%"
                params.extend([pat, pat])
            cur.execute(
                f"""
                SELECT title, grade, description,
                       population_age_min, population_age_max, population_sex,
                       source
                FROM guidelines
                WHERE grade IN ('A', 'B', 'I')
                  AND ({like_clause})
                ORDER BY grade, title
                """,
                params,
            )
            for row in cur.fetchall():
                if row["title"] in seen_titles:
                    continue
                recommendations.append(_row_to_rec(row))
                seen_titles.add(row["title"])

        conn.close()
        return {"recommendations": recommendations}
    except sqlite3.Error as e:
        return {"error": f"Database error: {e}"}
